### DISCUSSION TITLE
BUILD FAILED (macOS 14.5 on arm64 using ruby-build 20240530.1)

### DISCUSSION THREAD

**Discussion by megazoic** [rbenv/ruby-build#2404] [Build failures] [answered]
Hi: I am attempting to install ruby 3.2.4 or even 3.3.2 using

RUBY_CONFIGURE_OPTS=--with-libyamal-dir=/opt/local rbenv install 3.2.4

The build fails with 'Following extensions are not compiled psych:'

In the file .../ruby-3.2.4/ext/psych/mkmf.log I get...

package configuration for yaml-0.1 is not found
find_header: checking for yaml.h... -------------------- no

I have installed rbenv and ruby-build with macports and am installing a ruby without sudo on an unprivileged account.

I tried to install yaml with macports but that installs libyaml @0.2.5_0

I am running MacBook M3 chip and 14.5 os Sonoma

At https://trac.macports.org/ticket/69604 I see a comment 'It looks like rbenv is installing its own private copy of openssl first. If rbenv's intention is to be self-contained, then I guess it should either be installing its own private copy of libyaml or else telling ruby not to use libyaml, if that's possible. These would be bugs to report to the developers of rbenv. Or if they are not intending to be self-contained, then hopefully they have documentation explaining how to inform rbenv about the location of your copy of libyaml and anything else that's needed.'

So my question, how to I get rbenv to use the installed libyaml 0.2.5_0 instead of yaml-0.1 which it appears to be looking for? Or, what else should I do to be able to install a ruby with rbenv on this Mac using macports installation of rbenv and installing a ruby as a non privileged user?

Thank you.

**Comment by mislav** [comment, marked as answer]
> RUBY_CONFIGURE_OPTS=--with-libyamal-dir=/opt/local rbenv install 3.2.4

This looks almost correct, except that I see a small misspelling: it should be `--with-libyaml-dir`.

> It looks like rbenv is installing its own private copy of openssl first. If rbenv's intention is to be self-contained, then I guess it should either be installing its own private copy of libyaml or else telling ruby not to use libyaml, if that's possible.

This comment is correct in the fact that ruby-build can install a private copy of openssl, but it doesn't do that to be self-contained. It does that only on systems where it can't detect a suitable existing installation of openssl, which is most commonly macOS with its broken system OpenSSL/LibreSSL. Other external dependencies like libyaml or readline are typically optional, but a Ruby installation without openssl is unusable because it cannot connect to the Internet or install any gems.

rbenv/ruby-build doesn't have a way to be told *not* to use a system libyaml, but note that ruby-build will only pass a default `--with-libyaml-dir` flag if it detects either Homebrew-installed libyaml or a FreeBSD system: https://github.com/rbenv/ruby-build/blob/076d1a23169bb715942448d368be42424809d1e6/bin/ruby-build#L674-L676

In other cases, unless you pass `--with-libyaml-dir` yourself, this flag won't be sent while building Ruby. It will be up to ruby's own build system to determine how it's going to compile its yaml extension (called "psych").

**Reply by megazoic** [reply]
Thank you for catching that misspelling and for your description of ruby-build's use of libyaml. I realize that Homebrew has become very popular but would put a plug in for macports using an entirely separate (from Mac OS that is) environment.

**Reply by smanzini** [reply]
> > RUBY_CONFIGURE_OPTS=--with-libyamal-dir=/opt/local rbenv install 3.2.4
> 
> This looks almost correct, except that I see a small misspelling: it should be `--with-libyaml-dir`.
> 
> > It looks like rbenv is installing its own private copy of openssl first. If rbenv's intention is to be self-contained, then I guess it should either be installing its own private copy of libyaml or else telling ruby not to use libyaml, if that's possible.
> 
> This comment is correct in the fact that ruby-build can install a private copy of openssl, but it doesn't do that to be self-contained. It does that only on systems where it can't detect a suitable existing installation of openssl, which is most commonly macOS with its broken system OpenSSL/LibreSSL. Other external dependencies like libyaml or readline are typically optional, but a Ruby installation without openssl is unusable because it cannot connect to the Internet or install any gems.
> 
> rbenv/ruby-build doesn't have a way to be told _not_ to use a system libyaml, but note that ruby-build will only pass a default `--with-libyaml-dir` flag if it detects either Homebrew-installed libyaml or a FreeBSD system:
> 
> https://github.com/rbenv/ruby-build/blob/076d1a23169bb715942448d368be42424809d1e6/bin/ruby-build#L674-L676
> 
> In other cases, unless you pass `--with-libyaml-dir` yourself, this flag won't be sent while building Ruby. It will be up to ruby's own build system to determine how it's going to compile its yaml extension (called "psych").

hi @mislav I keep on receiving the build error BUILD FAILED (macOS 14.5 on arm64 using ruby-build 20240917) even using RUBY_CONFIGURE_OPTS=--with-libyaml-dir=/opt/local rbenv install 3.2.4 
can you help me?

**Reply by mislav** [reply]
@smanzini Hi, unfortunately without more info, I can't tell what's causing the failure. If you post part of the build log where there are some error message, it might be possible to determine what went wrong. Note that if the failure is when building a Ruby extension, such as "openssl" or "psych", more log information will be in a separate file for that extension, and the path to that log file will be mentioned in the main build log.

**Comment by bonol** [comment]
@mislav Hi, I have a build failing here with the same command. I have also attach the info of the build.
`RUBY_CONFIGURE_OPTS=--with-libyaml-dir=/opt/local rbenv install 3.2.4`

`BUILD FAILED (macOS 15.1 on arm64 using ruby-build 20241105)`

`readline.c:1903:37: error: use of undeclared identifier 'username_completion_function'; did you mean 'rl_username_completion_function'?
 1903 |                                     rl_username_completion_function);
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                     rl_username_completion_function
readline.c:79:42: note: expanded from macro 'rl_username_completion_function'
   79 | # define rl_username_completion_function username_completion_function
      |                                          ^
/usr/local/opt/readline/include/readline/readline.h:494:14: note: 'rl_username_completion_function' declared here
  494 | extern char *rl_username_completion_function (const char *, int);
      |              ^
compiling limits.c
1 error generated.
make[2]: *** [readline.o] Error 1
make[1]: *** [ext/readline/all] Error 2
make[1]: *** Waiting for unfinished jobs....
compiling ripper.c
checking ../.././parse.y and ../.././ext/ripper/eventids2.c
ld: warning: ignoring duplicate libraries: '-lffi', '-lruby.3.2'
linking shared-object rbconfig/sizeof.bundle
ripper.c:10310:9: warning: variable 'yynerrs' set but not used [-Wunused-but-set-variable]
 10310 |     int yynerrs = 0;
       |         ^
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
1 warning generated.
linking shared-object ripper.bundle
ld: warning: ignoring duplicate libraries: '-lruby.3.2'
make: *** [build-ext] Error 2
external command failed with status 2`

**Reply by mislav** [reply]
Hi, it's hard to tell what's going on here. You've used `RUBY_CONFIGURE_OPTS=--with-libyaml-dir=/opt/local`, but you didn't tell us why you've passed that option. Is libyaml on your system installed under `/opt/local`? That's pretty non-standard for macOS, is neither Apple nor Homebrew places it under that location by default. Can you confirm that your libyaml is there?

Second, the error message references `/usr/local/opt/readline/include/readline/readline.h`, which is also strange, since neither Apple nor Homebrew will ever put readline there.

In summary, your system appears to be customized in ways that most macOS systems are not, so without you sharing more information about that, there's little this forum can do to help. You could try using Homebrew to install readline and libyaml and allowing ruby-build to link to those, which should happen automatically.

**Reply by bonol** [reply]
> Hi, it's hard to tell what's going on here. You've used `RUBY_CONFIGURE_OPTS=--with-libyaml-dir=/opt/local`, but you didn't tell us why you've passed that option. Is libyaml on your system installed under `/opt/local`? That's pretty non-standard for macOS, is neither Apple nor Homebrew places it under that location by default. Can you confirm that your libyaml is there?
> 
> Second, the error message references `/usr/local/opt/readline/include/readline/readline.h`, which is also strange, since neither Apple nor Homebrew will ever put readline there.
> 
> In summary, your system appears to be customized in ways that most macOS systems are not, so without you sharing more information about that, there's little this forum can do to help. You could try using Homebrew to install readline and libyaml and allowing ruby-build to link to those, which should happen automatically.

Thanks so much for pointing all these out. Yap, there is nothing wrong with ruby-build. I found the root cause is my terminal setup has a mix of x86 and arm brew installed that confused ruby-build. I have reset the whole terminal with one version of brew and it's working fine now.

**Reply by rubywarlock** [reply]
I encounter the same error when using asdf.
hard: MacBook Pro 3 max
MacOS: 15.2 (24C101)
```
❯ asdf install ruby 3.0.7
==> Downloading openssl-1.1.1w.tar.gz...
-> curl -q -fL -o openssl-1.1.1w.tar.gz https://dqw8nmjcqpjn7.cloudfront.net/cf3098950cb4d853ad95c0841f1f9c6d3dc102dccfcacd521d93925208b76ac8
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 9661k  100 9661k    0     0  3616k      0  0:00:02  0:00:02 --:--:-- 3617k
==> Installing openssl-1.1.1w...
-> ./config "--prefix=$HOME/.asdf/installs/ruby/3.0.7/openssl" "--openssldir=$HOME/.asdf/installs/ruby/3.0.7/openssl/ssl" --libdir=lib zlib-dynamic no-ssl3 shared
-> make -j 14
-> make install_sw install_ssldirs
==> Installed openssl-1.1.1w to /Users/vitaly/.asdf/installs/ruby/3.0.7
==> Downloading ruby-3.0.7.tar.gz...
-> curl -q -fL -o ruby-3.0.7.tar.gz https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.7.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 20.2M  100 20.2M    0     0  2338k      0  0:00:08  0:00:08 --:--:-- 5024k
==> Installing ruby-3.0.7...

WARNING: ruby-3.0.7 is past its end of life and is now unsupported.
It no longer receives bug fixes or critical security updates.

ruby-build: using readline from homebrew
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew
-> ./configure "--prefix=$HOME/.asdf/installs/ruby/3.0.7" "--with-openssl-dir=$HOME/.asdf/installs/ruby/3.0.7/openssl" --enable-shared --with-readline-dir=/opt/homebrew/opt/readline --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-ext=openssl,psych,+
-> make -j 14

BUILD FAILED (macOS 15.2 on arm64 using ruby-build 20241225.2)

You can inspect the build directory at /var/folders/wt/vsk5xdg532x_4bsgdhnvy97h0000gn/T/ruby-build.20250112204255.21281.CqhtiZ
See the full build log at /var/folders/wt/vsk5xdg532x_4bsgdhnvy97h0000gn/T/ruby-build.20250112204255.21281.log
```

**Reply by mislav** [reply]
@rubywarlock Hi, it's hard to tell whether your failure is the same as others in this thread, since you haven't actually shared the build log. You could search for keywords like `error:` in the build log to check what went wrong. Since everything else in the output looks fine, and since you're installing Ruby 3.0 on macOS 15, I would guess that the problem is with "bigdecimal" and that you might get around it by applying the patch mentioned here: https://bugs.ruby-lang.org/issues/20760#note-4

**Comment by ahmaziz12** [comment]
I am using macbook 13 pro and macOS 15.6, I wanted to install ruby 2.6.6 using rbenv it gave the same error BUILD FAILED. I tried many commands but didnt work then my senior gave the below command and it worked.
`RUBY_CFLAGS=-DUSE_FFI_CLOSURE_ALLOC rbenv install`
